### PR TITLE
Add higher-order QAOA (HUBO) support

### PR DESCRIPTION
**Pull request by ahao27**
(This PR was needed to extend our type system to allow `Dict[Vector[UInt], Float]`, so the newer version of PR's status is just as follows.)
## Summary
- Add HUBO support to QAOA by decomposing higher-order Z terms with `phase_gadget` (CX ladder + RZ).
- Extend `QAOAConverter` to auto-dispatch between the quadratic fast path and the HUBO transpilation path.
- Refactor and rename QAOA algorithm primitives for consistency (`ising_cost`, `x_mixer`, `qaoa_layers`) and expose HUBO helpers in the public algorithm API.
- Add dedicated algorithm/optimization tests for phase-gadget behavior, HUBO transpilation, and quadratic regression.

## Test Plan
- Covered by CI (unit tests + docs tests).

<details>
<summary>
PR, first edition
</summary>

## Summary
- Add `apply_phase_gadget` for decomposing k-body Z-rotations into CX ladder + RZ primitives
- Extend `QAOAConverter` with HUBO path that auto-dispatches based on `spin_model.higher`
- Compose `ising_cost_circuit` in the HUBO cost circuit to avoid code duplication
- Export `apply_phase_gadget` in `qamomile.circuit.algorithm` public API
- Add comprehensive test suite (19 tests) with parametrized, random, and edge-case coverage

## File changes

| File | Change | Description |
|------|--------|-------------|
| `qamomile/circuit/algorithm/qaoa.py` | Modified | Add `apply_phase_gadget()` — decomposes k-body Z-rotations into CX ladder + RZ primitives (k=0: noop, k=1: RZ, k=2: RZZ, k≥3: CX ladder) |
| `qamomile/circuit/algorithm/__init__.py` | Modified | Export `apply_phase_gadget` in imports and `__all__` |
| `qamomile/optimization/qaoa.py` | Modified | Add class/method docstrings; add `_transpile_hubo()` HUBO path with phase-gadget decomposition; refactor `transpile()` to dispatch quadratic vs HUBO; compose `ising_cost_circuit` to eliminate duplication |
| `tests/optimization/test_qaoa_hubo.py` | Added | 19 tests: Hamiltonian construction (parametrized, random), transpilation (parametrized), phase gadget branches (k=0–3), edge cases (near-zero filtering, duplicate index accumulation, quadratic fast-path), end-to-end optimization |

## Test plan
- [x] `\local-review new_qamomile`
- [x] `uv run pytest tests/optimization/test_qaoa_hubo.py -v`
- [x] `uv run pytest tests/optimization/test_qaoa_decode.py -v`

</details>

### CODE REVIEW

File: tests/optimization/test_qaoa_hubo.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        sampleset = converter.decode(result)
+        argmin_index = np.argmin(sampleset.energy)
+        best_energy = sampleset.energy[argmin_index]
+        best_occurrence = sampleset.num_occurrences[argmin_index]
+        return best_energy * best_occurrence
+
+    x0 = [2.6, 0.11, 0.52, 0.45]
+    bounds = [(0, np.pi), (0, np.pi), (0, np.pi / 2), (0, np.pi / 2)]
+    res = scipy.optimize.minimize(
+        _obj, x0=x0, bounds=bounds, method="COBYLA", options={"maxiter": 100}
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply tests/optimization/test_qaoa_hubo.py]: Never mind. This is not a big deal. I'll also review soon.

File: qamomile/circuit/algorithm/qaoa.py [hunk 2]
```diff
@@ -1,13 +1,76 @@
+from __future__ import annotations
+
 import qamomile.circuit as qmc
 
 
+def apply_phase_gadget(
```
Comment (ksk-jij) [inline qamomile/circuit/algorithm/qaoa.py]: [my opinion]
Should be `@qkernel`.

File: qamomile/circuit/algorithm/qaoa.py [hunk 3]
```diff
@@ -1,13 +1,76 @@
...
+    if k == 0:
+        return q
+    if k == 1:
+        q[indices[0]] = qmc.rz(q[indices[0]], angle=angle)
+        return q
+    if k == 2:
+        q[indices[0]], q[indices[1]] = qmc.rzz(
+            q[indices[0]], q[indices[1]], angle=angle
+        )
+        return q
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply qamomile/circuit/algorithm/qaoa.py]: If we remove this, `for step in qmc.range(k - 1):` below will work directly.

* `k = 0`: error is raised
* `k = 1`: no for-statement works since `k-1 = 0` and just get `rz` as it should.
* `k = 2`: the first and second qubits will have `cx` and the second qubit will have `rz`. This is different from the current implementation, but still equivalent to `rzz`.

File: qamomile/circuit/algorithm/qaoa.py (hunk 2 above)
Comment (ksk-jij) [inline qamomile/circuit/algorithm/qaoa.py]: [my opinion]
It can go to `basic.py` since this is quite useful for anything. Also, I think `phase_gadget` is better than `apply_phase_gadget` if we can make it a proper `QKernel`.

File: qamomile/circuit/algorithm/qaoa.py [hunk 4]
```diff
@@ -69,11 +123,154 @@ def qaoa_state(
...
+    # CX forward ladder
+    for step in qmc.range(last):
+        next_step = step + 1
+        left = indices[step]
+        right = indices[next_step]
+        q[left], q[right] = qmc.cx(q[left], q[right])
+
+    # Apply RZ on the last qubit
+    q[indices[last]] = qmc.rz(q[indices[last]], angle=angle)
+
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply qamomile/circuit/algorithm/qaoa.py]: True. We will handle this in `BinaryModel`.

**Comment by ksk-jij** [comment]
Looks good to copilot, codex and me.
